libats-http/src/main/scala/com/advancedtelematic/libats/http/ErrorHandler.scala

```diff
@@ -16,46 +16,60 @@ import com.advancedtelematic.libats.data.{ErrorCode, ErrorCodes, ErrorRepresenta
 import de.heikoseeberger.akkahttpcirce.FailFastCirceSupport._
 import io.circe.syntax._
 import com.advancedtelematic.libats.codecs.CirceUuid._
-import io.circe.Json
+import io.circe.{Encoder, Json}
+import cats.syntax.option._
 
 import scala.reflect.ClassTag
 import scala.util.control.NoStackTrace
+import scala.language.existentials
 
 object Errors {
   import Directives._
   import ErrorRepresentation._
 
-  import scala.language.existentials
+  abstract class Error[T](val code: ErrorCode,
+                          val responseCode: StatusCode,
+                          val msg: String,
+                          val cause: Throwable = null,
+                          val errorId: UUID = UUID.randomUUID()) extends Throwable(msg, cause) with NoStackTrace
 
-  trait Error[T] extends NoStackTrace {
-    implicit val ct: ClassTag[T]
-    def name = ct.runtimeClass.getSimpleName
-  }
-
-  case class MissingEntity[T]()(implicit val ct: ClassTag[T]) extends Throwable(s"Missing entity: ${ct.runtimeClass.getSimpleName}") with Error[T]
-  case class EntityAlreadyExists[T]()(implicit val ct: ClassTag[T]) extends Throwable(s"Entity already exists: ${ct.runtimeClass.getSimpleName}") with Error[T]
+  case class JsonError(code: ErrorCode,
+                       responseCode: StatusCode,
+                       json: Json,
+                       msg: String,
+                       errorId: UUID = UUID.randomUUID()) extends Throwable(msg) with NoStackTrace
 
   case class RawError(code: ErrorCode,
                       responseCode: StatusCode,
-                      desc: String) extends Exception(desc) with NoStackTrace
+                      desc: String,
+                      errorId: UUID = UUID.randomUUID()) extends Exception(desc) with NoStackTrace
+
+  case class MissingEntity[T]()(implicit ct: ClassTag[T]) extends
+    Error[T](ErrorCodes.MissingEntity, StatusCodes.NotFound, s"Missing entity: ${ct.runtimeClass.getSimpleName}")
+
+  case class EntityAlreadyExists[T]()(implicit ct: ClassTag[T]) extends
+    Error[T](ErrorCodes.ConflictingEntity, StatusCodes.Conflict, s"Entity already exists: ${ct.runtimeClass.getSimpleName}")
+
+  def RemoteServiceError(msg: String, cause: Option[Json] = None, errorId: UUID = UUID.randomUUID()) = {
+    val causeJson = cause.getOrElse(Json.obj())
+    JsonError(ErrorCodes.RemoteServiceError, StatusCodes.BadGateway, causeJson, msg, errorId)
```

Comment:
So it seems like we can easily forget to log the `errorId`, so if for example the `UUID` is created when calling this function then we will give a `UUID` to the user that we don't know anything about. I suppose we already are in that situation so this is not that bad.

Reply:
Also do we need to check if `isProd` for `RemoteServiceError`?

Reply:
What do you mean though? You don't always have a uuid so if you dont have one we create one here?

Information on why the service failed is useful for any client in any environment. The error itself (cause) should have been cleanup with `isProd` in the remote service.

Reply:
What I mean with the uuid is that if it is created here and sent to the user but not stored (or logged) anywhere on the server the uuid is useless. But I suppose that is fine and that we in our services should try make sure to use the uuid when it makes sense.

Okay I suppose that is fine for the `isProd`, I'm just making sure I understand what is safe and what isn't to reveal.

Reply:
Yes, that is true, but I kept it like that for two reasons:

- I don't have a nice way to log it, we could use `LoggerFactory` on that object to get a logger but it's not ideal, but more importantly:

- Everything we would log in the server is already in the response, so there is no real point in logging it? I added all this so that we see the point of logging we can easily add logging with that uuid, probably the next iteration of this anyway we could probably just log what we are responding with.